### PR TITLE
Apply multiple bindings to the same subject

### PR DESCRIPTION
**Pull request by glyn**
Fixes https://github.com/knative/pkg/issues/1786.

cc @mattmoor @n3wscott

### CODE REVIEW

**Comment by mattmoor** [comment]
This generally seems reasonable to me.

File: webhook/psbinding/index.go [hunk 1]
```diff
@@ -79,3 +83,76 @@ func (im inexactMatcher) Get(key inexactKey, ls labels.Set) (bindable Bindable,
...
+func (idx *index) setIndex(exact exactMatcher, inexact inexactMatcher) {
+	idx.lock.Lock()
+	defer idx.lock.Unlock()
+	idx.exact = exact
+	idx.inexact = inexact
+}
+
+// LookUp returns the Bindables associated with a resource with the given group, kind, namespace, name, and labels.
+// TODO: find all the matching Bindables instead of just one.
+func (idx *index) LookUp(group string, kind string, namespace string, name string, labels labels.Set) []Bindable {
```
Comment (vagababov) [inline webhook/psbinding/index.go:134]:
```suggestion
func (idx *index) LookUp(group, kind, namespace, name string, labels labels.Set) []Bindable {
```

File: webhook/psbinding/index.go [hunk 2]
```diff
@@ -79,3 +83,76 @@ func (im inexactMatcher) Get(key inexactKey, ls labels.Set) (bindable Bindable,
...
+// index is a collection of Bindables indexed by their subject resources.
+type index struct {
+	// lock protects access to exact and inexact
+	lock    sync.RWMutex
+	exact   exactMatcher
+	inexact inexactMatcher
+}
+
+// IndexBuilder allows an index to be built atomically
+type IndexBuilder struct {
```
Comment (vagababov) [inline webhook/psbinding/index.go:96]: this and the next function seem to be local, so no need to export.

File: webhook/psbinding/index.go (hunk 1 above)
Comment (glyn) [reply webhook/psbinding/index.go]: @vagababov Agreed. What I actually did was group the first four parameters into an `exactKey` which makes the call site more readable.

File: webhook/psbinding/index.go (hunk 2 above)
Comment (glyn) [reply webhook/psbinding/index.go:96]: @vagababov Yes. My reasoning was that the matcher methods in `index.go` were exported and so I tried to adopt the same convention. (A lint complains about unexported `indexBuilder` being returned from `NewIndexBuilder`, so exporting the latter leads to exporting the former).

Maybe the best convention is to just unexport everything in `index.go`.

Comment (glyn) [reply webhook/psbinding/index.go]: @vagababov Fixed. Please see the second commit.

Comment (vagababov) [reply webhook/psbinding/index.go]: yeah, my point was that `NewIndexBuilder` should not be exported either 😄

File: webhook/psbinding/index.go [hunk 3]
```diff
@@ -79,3 +83,71 @@ func (im inexactMatcher) Get(key inexactKey, ls labels.Set) (bindable Bindable,
...
+	// Next look for inexact matches.
+	if sb, ok := idx.inexact.get(inexactKey{
+		Group:     key.Group,
+		Kind:      key.Kind,
+		Namespace: key.Namespace,
+	}, labels); ok {
+		return []Bindable{sb}
+	}
+
+	return []Bindable{}
```
Comment (vagababov) [inline webhook/psbinding/index.go:152]: just nil here?

**Review by vagababov — COMMENTED** [review]
Looks good.
Thanks.

File: webhook/psbinding/index.go (hunk 2 above)
Comment (glyn) [reply webhook/psbinding/index.go]: Sure. That's addressed now.

File: webhook/psbinding/index.go (hunk 3 above)
Comment (glyn) [reply webhook/psbinding/index.go]: Fixed. Clearly I've been writing too much Rust code...

**Comment by glyn** [comment]
@vagababov Hi. The PR is functionally complete. I'm going to beef up the exact and inexact reverse index tests. Meanwhile, PTAL.

**Comment by glyn** [comment]
Note: the coverage of `psbinding.go` has decreased only because code was factored out and the uncovered lines are now a larger proportion of the total.

**Comment by glyn** [comment]
@scothis PTAL and see if this fixes the downstream problem.

**Comment by glyn** [comment]
>  I'm going to beef up the exact and inexact reverse index tests.

Done.

File: webhook/psbinding/psbinding.go [hunk 4]
```diff
@@ -197,56 +193,48 @@ func (ac *Reconciler) Admit(ctx context.Context, request *admissionv1.AdmissionR
...
-	}()
-	if fb == nil {
+	// Look up the Bindables for this resource.
+	fbs := ac.index.lookUp(exactKey{
+		Group:     request.Kind.Group,
+		Kind:      request.Kind.Kind,
+		Namespace: request.Namespace,
+		Name:      orig.Name,
+	},
+		labels.Set(orig.Labels))
```
Comment (scothis) [inline webhook/psbinding/psbinding.go:203]: this formatting is a bit odd

**Review by scothis — COMMENTED** [review]
Tested this PR with the [vmware-labs/service-bindings](https://github.com/vmware-labs/service-bindings) repo, and now multiple bindings apply to a Job on creation as expected. Thanks @glyn

/lgtm

File: webhook/psbinding/psbinding.go (hunk 4 above)
Comment (glyn) [reply webhook/psbinding/psbinding.go:203]: Agreed. Note that the same oddity occurred in the deleted code just above. Thanks `go fmt`.

Comment (vagababov) [reply webhook/psbinding/psbinding.go:203]: I'd join with the previous line to avoid this.

Comment (glyn) [reply webhook/psbinding/psbinding.go:202]: Fixed two occurrence. Looks a little better now.

Comment (glyn) [reply webhook/psbinding/psbinding.go:202]: Fixed.

Comment (glyn) [reply webhook/psbinding/psbinding.go:202]: Fixed.

**Pull request merged**